sql/core/src/main/scala/org/apache/spark/sql/execution/subquery.scala

```diff
@@ -131,14 +131,15 @@ case class InSubqueryExec(
 
   def updateResult(): Unit = {
     val rows = plan.executeCollect()
-    result = child.dataType match {
-      case _: StructType => rows.toArray
-      case _ => rows.map(_.get(0, child.dataType))
-    }
+    // The child is always partition column. So it must be AtomicType.
```

Comment:
Is this true? Seems `PlanSubqueries` will possibly pass a struct into?

https://github.com/apache/spark/blob/68034a805607ced50dbedca73dfc7eaf0102dde8/sql/core/src/main/scala/org/apache/spark/sql/execution/subquery.scala#L186-L196

Reply:
`InSubqueryExec` is for Dynamic Partition Pruning only:
https://github.com/apache/spark/blob/68034a805607ced50dbedca73dfc7eaf0102dde8/sql/core/src/main/scala/org/apache/spark/sql/execution/subquery.scala#L109-L117

`PlanSubqueries` actually has one element. It is from [PlanDynamicPruningFilters.scala#L88-L89](https://github.com/apache/spark/blob/68034a805607ced50dbedca73dfc7eaf0102dde8/sql/core/src/main/scala/org/apache/spark/sql/dynamicpruning/PlanDynamicPruningFilters.scala#L88-L89).

I'm thinking could we move [subquery.scala#L185-L196](https://github.com/apache/spark/blob/68034a805607ced50dbedca73dfc7eaf0102dde8/sql/core/src/main/scala/org/apache/spark/sql/execution/subquery.scala#L185-L196) to [PlanDynamicPruningFilters.scala#L88-L89](https://github.com/apache/spark/blob/68034a805607ced50dbedca73dfc7eaf0102dde8/sql/core/src/main/scala/org/apache/spark/sql/dynamicpruning/PlanDynamicPruningFilters.scala#L88-L89) to make logic more clear.

Reply:
cc @maryannxue

Reply:
I don't get it. What it meant "PlanSubqueries actually has one element."? `PlanSubqueries` has the code to pass a struct into `InSubqueryExec`, isn't? Otherwise it is useless and we should remove it too.

Reply:
I mean there is always only one element:
![image](https://user-images.githubusercontent.com/5399861/90845541-d5386b00-e398-11ea-9c19-be2d2fdb3442.png)

See changes if we move subquery.scala#L185-L196 to PlanDynamicPruningFilters.scala#L88-L89: https://github.com/apache/spark/pull/29475/commits/3e98608c3afae00599cdb1fa1d6156df168a6ffa

Reply:
Yeah, I meant https://github.com/apache/spark/blob/68034a805607ced50dbedca73dfc7eaf0102dde8/sql/core/src/main/scala/org/apache/spark/sql/execution/subquery.scala#L186-L196, especially L189-L193 that creates `CreateNamedStruct`.

If "The child is always partition column" is true, then it means we don't need the code at `subquery.scala` that passes a struct into `InSubqueryExec`. Isn't?

Reply:
Yes.